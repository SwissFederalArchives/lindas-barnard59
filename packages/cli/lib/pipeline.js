import { dirname, resolve } from 'path'
import fromStream from 'rdf-dataset-ext/fromStream.js'
import rdf from '@zazuko/env'
import { isGraphPointer } from 'is-graph-pointer'
import fromFile from 'rdf-utils-fs/fromFile.js'
import findPipeline from '../findPipeline.js'
import discoverManifests from './discoverManifests.js'
import ns from './namespaces.js'

const discoverOperations = async () => {
  const ops = rdf.termMap()
  for await (const { manifest } of discoverManifests()) {
    manifest
      .has(rdf.ns.rdf.type, ns.p.Operation)
      .forEach(operation => {
        const impl = operation.out(ns.code.implementedBy)
        const type = impl.out(ns.rdf.type).term
        const link = impl.out(ns.code.link).term
        ops.set(operation.term, { type, link })
      })
  }

  return ops
}

export const desugar = async (dataset, { logger, knownOperations } = {}) => {
  knownOperations = knownOperations ?? await discoverOperations()
  const ptr = rdf.clownface({ dataset })
  ptr.has(ns.p.stepList).out(ns.p.stepList).forEach(listPointer => {
    for (const step of listPointer.list()) {
      if (isGraphPointer(step.has(ns.rdf.type, ns.p.Step)) ||
          isGraphPointer(step.has(ns.rdf.type, ns.p.Pipeline))) {
        continue
      }
      // we expect a known operation
      const [quad] = step.dataset.match(step.term)
      const knownStep = knownOperations.get(quad?.predicate)
      if (!knownStep) {
        logger?.warn(`Operation <${quad?.predicate.value}> not found in known manifests. Have you added the right \`branard59-*\` package as dependency?`)
        continue
      }

      const { type, link } = knownStep
      const args = step.out(quad.predicate)
      step.deleteOut(quad.predicate)
      // keep args only if non-empty
      if (!ns.rdf.nil.equals(args.term)) {
        step.addOut(ns.code.arguments, args)
      }
      step.addOut(ns.rdf.type, ns.p.Step)
      step.addOut(ns.code.implementedBy, moduleNode => {
        moduleNode.addOut(ns.rdf.type, type)
        moduleNode.addOut(ns.code.link, link)
      })
    }
  })

  return ptr.dataset
}

async function fileToDataset(filename) {
  return fromStream(rdf.dataset(), fromFile(filename))
}

export async function parse(filename, iri, { logger } = {}) {
  const dataset = await fileToDataset(filename)
  const ptr = findPipeline(await desugar(dataset, { logger }), iri)

  return {
    basePath: resolve(dirname(filename)),
    ptr,
  }
}
