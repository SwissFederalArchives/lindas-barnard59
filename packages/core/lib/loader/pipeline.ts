import type { LoaderRegistry } from 'rdf-loaders-registry'
import type { Environment } from 'lindas-barnard59-env'
import type { GraphPointer } from 'clownface'
import createPipeline from '../factory/pipeline.js'
import type { Context, VariableMap } from '../../index.js'

interface Options {
  basePath: string
  context: Context
  loaderRegistry: LoaderRegistry
  variables: VariableMap
}

async function loader(ptr: GraphPointer, { basePath, context, loaderRegistry, variables }: Options) {
  if (ptr.has(context.env.ns.rdf.type, context.env.ns.p.Pipeline).terms.length > 0) {
    return createPipeline(ptr, { basePath, context, loaderRegistry, logger: context.logger, variables }).stream
  }

  throw new Error('Unrecognized or missing pipeline type')
}

loader.register = (registry: LoaderRegistry, env: Environment) => {
  registry.registerNodeLoader(env.ns.p.Pipeline, loader)
}

export default loader
