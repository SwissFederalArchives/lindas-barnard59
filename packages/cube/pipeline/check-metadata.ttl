@prefix code: <https://code.described.at/> .
@prefix p: <https://pipeline.described.at/> .
@prefix shacl: <https://lindas-barnard59.zazuko.com/operations/shacl/> .
@prefix base: <https://lindas-barnard59.zazuko.com/operations/base/> .
@prefix n3: <https://lindas-barnard59.zazuko.com/operations/formats/n3/> .
@prefix ntriples: <https://lindas-barnard59.zazuko.com/operations/formats/ntriples/> .
@prefix rdf: <https://lindas-barnard59.zazuko.com/operations/rdf/> .
@prefix rdfs: <http://www.w3.org/2000/01/rdf-schema#> .

@base <http://barnard59.zazuko.com/pipeline/cube-validation/> .


_:profile a p:Variable ;
  p:name "profile" ;
  rdfs:label "cube constraint profile URL" ;
  p:value "https://cube.link/latest/shape/standalone-constraint-constraint" ;
.

_:profileFormat a p:Variable ;
  p:name "profileFormat" ;
  rdfs:label "Override the profile format" ;
  p:required false ;
.

<check-metadata> a p:Pipeline , p:Readable ;
  p:variables [ p:variable _:profile, _:profileFormat ] ;
  p:steps
    [
      p:stepList (
        [ base:stdin () ]
        [ n3:parse () ]
        [ rdf:getDataset () ]
        [ shacl:report (_:getProfile) ]
        [ base:flatten () ]
        [ ntriples:serialize () ]
      )
    ]
.

_:getProfile a p:Pipeline , p:ReadableObjectMode;
  p:steps
    [
      p:stepList
        (
          [ rdf:open ( "profile"^^p:VariableName "profileFormat"^^p:VariableName ) ]
          [ rdf:transformCodeImports () ]
        )
    ]
  .
