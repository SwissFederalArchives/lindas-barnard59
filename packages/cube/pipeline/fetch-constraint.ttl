@prefix code: <https://code.described.at/> .
@prefix p: <https://pipeline.described.at/> .
@prefix ntriples: <https://lindas-barnard59.zazuko.com/operations/formats/ntriples/> .
@prefix rdfs: <http://www.w3.org/2000/01/rdf-schema#> .
@prefix : <http://barnard59.zazuko.com/pipeline/cube-validation/> .

[
  code:imports <./query-constraint> ;
  code:extension "ttl" ;
] .

_:endpoint a p:Variable ;
  p:name "endpoint" ;
  rdfs:label "SPARQL endpoint" ;
.

_:cube a p:Variable ;
  p:name "cube" ;
  rdfs:label "cube URI" ;
.

:fetch-constraint a p:Pipeline , p:Readable ;
  p:variables [ p:variable _:endpoint, _:cube ] ;
  p:steps
    [
      p:stepList
        (
          :queryConstraint
          [ ntriples:serialize () ]
        )
    ]
.
