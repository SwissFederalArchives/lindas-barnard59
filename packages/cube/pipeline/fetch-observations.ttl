@prefix code: <https://code.described.at/> .
@prefix p: <https://pipeline.described.at/> .
@prefix sparql: <https://barnard59.zazuko.com/operations/sparql/raw/> .
@prefix ntriples: <https://barnard59.zazuko.com/operations/formats/ntriples/> .
@prefix rdfs: <http://www.w3.org/2000/01/rdf-schema#> .

@base <http://barnard59.zazuko.com/pipeline/cube-validation/> .

_:endpoint a p:Variable ;
  p:name "endpoint" ;
  rdfs:label "SPARQL endpoint" ;
.

_:cube a p:Variable ;
  p:name "cube" ;
  rdfs:label "cube URI" ;
.

<fetch-observations> a p:Pipeline , p:Readable ;
  p:variables [ p:variable _:endpoint, _:cube ] ;
  p:steps
    [
      p:stepList
        (
          _:queryObservations
          # [ ntriples:serialize () ]
        )
    ]
.


_:queryObservations sparql:construct
    [
      code:name "endpoint";
      code:value "endpoint"^^p:VariableName
    ] ,
    [
      code:name "query";
      code:value """
      PREFIX cube: <https://cube.link/>

      CONSTRUCT { ?s ?p ?o }
      WHERE {
        <${cube}> cube:observationSet/cube:observation ?s .
        ?s ?p ?o
      }
      """^^code:EcmaScriptTemplateLiteral
    ]
.
