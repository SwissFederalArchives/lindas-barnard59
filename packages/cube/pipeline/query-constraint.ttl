@prefix code: <https://code.described.at/> .
@prefix p: <https://pipeline.described.at/> .
@prefix sparql: <https://lindas-barnard59.zazuko.com/operations/sparql/> .
@prefix : <http://barnard59.zazuko.com/pipeline/cube-validation/> .

# relying on cbd is not ideal (vendor specific)
:queryConstraint a p:Pipeline, p:ReadableObjectMode ;
  p:steps
    [
      p:stepList
        (
          [
            sparql:construct
              [ code:name "endpoint"; code:value "endpoint"^^p:VariableName ] ,
              [
                code:name "query";
                code:value """
                #pragma describe.strategy cbd

                PREFIX cube: <https://cube.link/>

                DESCRIBE ?s
                WHERE {
                  <${cube}> cube:observationConstraint ?s .
                }
                """^^code:EcmaScriptTemplateLiteral
              ]
          ]
        )
    ]
  .
  