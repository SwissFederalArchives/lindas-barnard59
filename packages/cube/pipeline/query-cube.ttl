@prefix code: <https://code.described.at/> .
@prefix p: <https://pipeline.described.at/> .
@prefix sparql: <https://lindas-barnard59.zazuko.com/operations/sparql/> .
@prefix : <http://barnard59.zazuko.com/pipeline/cube-validation/> .

# relying on cbd is not ideal (vendor specific)
:queryCube a p:Pipeline , p:ReadableObjectMode ;
  p:steps
    [
      p:stepList
        (
          [
            sparql:construct
              [ code:name "endpoint"; code:value "endpoint"^^p:VariableName ] ,
              [
                code:name "query";
                code:value """
                #pragma describe.strategy cbd

                DESCRIBE <${cube}>
                """^^code:EcmaScriptTemplateLiteral
              ]
          ]
        )
    ]
  .
