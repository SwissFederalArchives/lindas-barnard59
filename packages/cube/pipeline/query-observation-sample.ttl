@prefix code: <https://code.described.at/> .
@prefix p: <https://pipeline.described.at/> .
@prefix sparql: <https://lindas-barnard59.zazuko.com/operations/sparql/> .
@prefix : <http://barnard59.zazuko.com/pipeline/cube-validation/> .

:queryObservationSample a p:Pipeline, p:ReadableObjectMode ;
  p:steps
    [
      p:stepList
        (
          [
            sparql:construct
              [ code:name "endpoint"; code:value "endpoint"^^p:VariableName ] ,
              [
                code:name "query";
                code:value """
                PREFIX cube: <https://cube.link/>
                
                CONSTRUCT { 
                  <${cube}> cube:observationSet ?set .
                  ?set cube:observation ?obs .
                  ?obs ?p ?o .
                } 
                WHERE {
                  { SELECT ?set ?obs WHERE {
                      <${cube}> cube:observationSet ?set .
                      ?set cube:observation ?obs .
                    } LIMIT 1
                  }
                  ?obs ?p ?o
                }
                """^^code:EcmaScriptTemplateLiteral
              ]
          ]
        )
    ]
  .
