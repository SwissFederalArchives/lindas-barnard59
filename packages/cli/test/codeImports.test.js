import { createRequire } from 'node:module'
import env from 'lindas-barnard59-env'
import formatsPretty from '@rdfjs-elements/formats-pretty'
import chai, { expect } from 'chai'
import { jestSnapshotPlugin } from 'mocha-chai-jest-snapshot'
import * as pipeline from '../lib/pipeline.js'

const require = createRequire(import.meta.url)
env.formats.import({
  serializers: formatsPretty.serializers,
})

const prefixes = [
  'sh',
  'rdf',
  ['p', 'https://pipeline.described.at/'],
  ['code', 'https://code.described.at/'],
]

describe('pipeline', () => {
  chai.use(jestSnapshotPlugin())

  context('with code:imports', () => {
    it('merges the pipelines', async () => {
      // given
      const filename = require.resolve('lindas-barnard59-test-e2e/definitions/code-imports/main.ttl')

      // when
      const { ptr } = await pipeline.parse(filename, env.namedNode('http://example.org/pipeline/'))

      // then
      const pretty = await ptr.dataset.serialize({ format: 'text/turtle', prefixes })
      expect(pretty).toMatchSnapshot()
    })
  })
})
