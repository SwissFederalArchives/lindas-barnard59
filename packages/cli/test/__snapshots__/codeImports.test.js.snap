// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`pipeline with code:imports merges the pipelines 1`] = `
"@prefix sh: <http://www.w3.org/ns/shacl#> .
@prefix rdf: <http://www.w3.org/1999/02/22-rdf-syntax-ns#> .
@prefix p: <https://pipeline.described.at/> .
@prefix code: <https://code.described.at/> .

<http://example.org/pipeline/> a p:Pipeline ;
	p:steps [
		p:stepList (
			[
				rdf:type p:Step ;
				code:arguments (
					\\"foo\\"
					\\"bar\\"
				) ;
				code:implementedBy [
					rdf:type code:EcmaScriptModule ;
					code:link <node:lindas-barnard59-base/Readable.js#from> ;
				] ;
			]
			<http://example.org/pipeline/out>
		) ;
	] .

<http://example.org/pipeline/out> a p:Pipeline, p:Writable ;
	p:steps [
		p:stepList (
			[
				rdf:type p:Step ;
				code:implementedBy [
					rdf:type code:EcmaScriptModule ;
					code:link <node:lindas-barnard59-base/process.js#stdout> ;
				] ;
			]
		) ;
	] .

"
`;
